src/resolve-info.ts

```diff
@@ -196,18 +231,31 @@ const memoizedExpandFieldNode = MEMOIZATION
 function expandFieldNode(
   schema: GraphQLSchema,
   fragments: FragmentsType,
-  node: FieldNode,
+  node: JitFieldNode,
   fieldType: GraphQLOutputType
 ): FieldExpansion | LeafField {
+  const shouldInclude = (variables: ShouldIncludeVariables): boolean => {
+    const expr = node.__internalShouldInclude as string;
+    // eslint-disable-next-line no-new-func
+    const fn = new Function(`__context`, `return ${expr}`);
+
+    return fn(variables);
+  };
+
   if (node.selectionSet == null) {
-    return createLeafField({});
+    return createLeafField({}, shouldInclude);
   }
 
   // there is a selectionSet which makes the fieldType a CompositeType
   const typ = memoizedResolveEndType(fieldType) as GraphQLCompositeType;
   const possibleTypes = memoizedGetPossibleTypes(schema, typ);
 
-  const fieldExpansion: FieldExpansion = {};
+  const fieldExpansion: FieldExpansion = Object.create(
+    {
+      __shouldInclude: shouldInclude
+    },
+    {}
+  );
```

Comment:
I have 0 ideas how, but these lines break objects merging and lead to the failing test in this PR.